### PR TITLE
Add XCom.clear so it's hookable in custom XCom backend

### PR DESCRIPTION
**Pull request by sighingnow**
Airflow supports custom XCom backend, but there are still a direct xcom query in `TaskInstance`, which breaks the custom XCom backend do proper pre/post processing when deleting values in XCom storage.

This pull requests add a new interface `XCom.clear` to `BaseXCom` and use it when clearing results for given dags and tasks, instead of directly querying and operating the underlying database.

The change in this pull request will allow the custom XCom backend process the data lifecycle easier, and is fully backwards compatible.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/xcom.py [hunk 1]
```diff
@@ -224,6 +224,52 @@ def delete(cls, xcoms, session=None):
             session.delete(xcom)
         session.commit()
 
+    @classmethod
+    @provide_session
+    def clear(
+        cls,
+        execution_date: pendulum.DateTime,
+        dag_id: Optional[str] = None,
+        task_id: Optional[str] = None,
```
Comment (ashb) [inline airflow/models/xcom.py]: Why are these optional?

File: airflow/models/xcom.py [hunk 2]
```diff
@@ -224,6 +224,52 @@ def delete(cls, xcoms, session=None):
...
+        include_prior_dates: bool = False,
```
Comment (ashb) [inline airflow/models/xcom.py]: Why did you add this flag?

File: airflow/models/xcom.py [hunk 3]
```diff
@@ -224,6 +224,52 @@ def delete(cls, xcoms, session=None):
...
+
+        if include_prior_dates:
+            filters.append(cls.execution_date <= execution_date)
+        else:
+            filters.append(cls.execution_date == execution_date)
+
+        session.query(cls).filter(
+            and_(*filters)
+        ).delete()
+        session.commit()
```
Comment (ashb) [inline airflow/models/xcom.py]: https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#database-session-handling

```suggestion
```

**Review by ashb — CHANGES_REQUESTED** [review]
Moving it to BaseXCom makes sense, but I'm not sure about the type signature you have chosen.

File: airflow/models/xcom.py (hunk 1 above)
Comment (sighingnow) [reply airflow/models/xcom.py]: Fixed.

File: airflow/models/xcom.py (hunk 2 above)
Comment (sighingnow) [reply airflow/models/xcom.py]: Fixed.

File: airflow/models/xcom.py (hunk 3 above)
Comment (sighingnow) [reply airflow/models/xcom.py]: Thanks for pointing to the hint, fixed.

**Comment by sighingnow** [comment]
> but I'm not sure about the type signature you have chosen.

I was imaging adding more flexible to the interface for possible future usage. I have revert those extensions to make sure it is fully backwards compatible.

**Comment by ashb** [comment]
> I was imaging adding more flexible to the interface for possible future usage

We can add this in the future, but lets do that _when_ we have a usecase/need for it :)

File: airflow/models/taskinstance.py [hunk 4]
```diff
@@ -656,11 +656,12 @@ def clear_xcom_data(self, session=None):
...
-            XCom.task_id == self.task_id,
-            XCom.execution_date == self.execution_date,
-        ).delete()
+        XCom.clear(
+            dag_id=self.dag_id,
+            task_id=self.task_id,
+            execution_date=self.execution_date,
+            session=session,
+        )
         session.commit()
```
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
```

While we are here, lets fix up this too.

File: airflow/models/xcom.py [hunk 5]
```diff
@@ -224,6 +224,33 @@ def delete(cls, xcoms, session=None):
...
+        dag_id: str,
+        task_id: str,
+        session: Session = None,
+    ) -> None:
+        """
+        Clears all XCom data from the database for the task instance
+
+        :param execution_date: Execution date for the task
+        :type execution_date: pendulum.datetime
+        :param dag_id: Only pulls XCom from this DAG.
```
Comment (ashb) [inline airflow/models/xcom.py]:
```suggestion
        :param dag_id: DAG ID to clear XCom of.
```

File: airflow/models/xcom.py [hunk 6]
```diff
@@ -224,6 +224,33 @@ def delete(cls, xcoms, session=None):
...
+        :type dag_id: str
+        :param task_id: Only XComs from task with matching id will be pulled.
```
Comment (ashb) [inline airflow/models/xcom.py]:
```suggestion
        :param task_id: Only XComs from task with matching id will be cleared.
```

File: airflow/models/xcom.py [hunk 7]
```diff
@@ -224,6 +224,33 @@ def delete(cls, xcoms, session=None):
...
+        dag_id: str,
+        task_id: str,
+        session: Session = None,
+    ) -> None:
+        """
+        Clears all XCom data from the database for the task instance
+
+        :param execution_date: Execution date for the task
+        :type execution_date: pendulum.datetime
+        :param dag_id: DAG ID to clear XCom of.
```
Comment (ashb) [inline airflow/models/xcom.py]:
```suggestion
        :param dag_id: ID of DAG to clear the XCom for.
```

**Review by ashb — APPROVED** [review]

**Comment by sighingnow** [comment]
Ash, thank you for the assistant!

**Comment by sighingnow** [comment]
Could we get this one landed? The failure on CI is not caused by this pull request.

**Comment by ephraimbuddy** [comment]
Please rebase @sighingnow

**Comment by sighingnow** [comment]
> Please rebase @sighingnow

Rebased to latest main.

**Comment by sighingnow** [comment]
> Please rebase @sighingnow

Hi @ephraimbuddy could we get this pull request merged? I think the failed CI settings is not caused by this pull request, as I have noticed similar random failure on other pull requests that need full test.

**Review by uranusjr — APPROVED** [review]

**Comment by MatrixManAtYrService** [comment]
I see that `XCom.clear` has parameters that let you know what you're clearing (`dag_id`, `task_id`, `execution_date`).  When I set out to implement a custom backend, I totally expected these--so I'm glad they're here ❤️ .

What confuses me is that [the recommended approach for custom xcom backends ](https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html#custom-backends) is to override `serialize_value` and `deserialize_value`, and those **don't** have parameters of that sort.

So when I'm "serializing" a value, I don't know which task it's for.  I have to do something like generate a UUID or a hash of the data or something to come up with a key that won't collide with other values (this key is then used to recover the stored value while "deserializing").   So if I later want to clear that value, I can't distinguish it from its neighbors which I shouldn't clear.

Am I missing something?  Or does inviting users to implement `clear` put them in a challenging place re: how they implement the other two functions?

**Comment by sighingnow** [comment]
Hi @MatrixManAtYrService, I would like to share our experiences of implementing the airflow XCom backend in vineyard (see also: https://github.com/v6d-io/v6d/blob/main/python/vineyard/contrib/airflow/xcom/backend.py)

Background: vineyard a in-memory storage that that share complex composable data objects with clients in a zero-copy fashion using memory sharing.

We implements a XCom backend based on vineyard. Each object in vineyard has own unique identifier (`ObjectID`), then

+ for serialization, we first put the data into vineyard, then store the returned object id into the db of airflow
+ for deserailization, we implemented as first retrieve the object id, then getting object from vineyard
+ for deletion, we first query the object id for required `(dag, task, execution_date)` then deleting it from vineyard.

Hope that could be helpful for you.

**Comment by MatrixManAtYrService** [comment]
That gives me the missing pieces @sighingnow, thank you.  I should've looked for more functions to override (like `set` and `delete`).

**Comment by jmaldon1** [comment]
@sighingnow @MatrixManAtYrService Looking for an explanation of `clear`.

When is the `clear` method called in the DAG process?
How are we avoiding the case that we delete the data from the external database, but the tasks fails and has to rerun?
If the task reruns, won't it need to query for the data again, but if we deleted it already, then it wont be there anymore.

**Comment by MatrixManAtYrService** [comment]
My impression is that `clear` is called on a task's xcoms before it runs. This is to prevent data from a prior rerun from influencing the current task execution. This way a task can assume that its XComs are always in a blank slate, whether it's a run or a rerun.

I'll have an opportunity to verify this impression soon and I'll be sure to report back and correct it if I'm wrong.  I'll also let you know if I stumble across an understanding of `delete`

**Comment by jmaldon1** [comment]
@MatrixManAtYrService So to get it straight:

(Clear Task A XCOM) Task A -> (Clear Task B XCOM) Task B -> (Clear Task C XCOM) Task C.

Now when the DAG completes, there will still be XCOMs for all tasks in the database, is there a way to delete those?

**Comment by sighingnow** [comment]
> @sighingnow @MatrixManAtYrService Wondering if you guys could help me understand:
> 
> When is the [`clear`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/xcom/index.html#airflow.models.xcom.BaseXCom.clear) method called in the DAG process? How are we avoiding the case that we clear the data from the external database, but the tasks fails and has to rerun? If the task reruns, won't it need to query for the data again, but if we deleted it already, then it wont be there anymore.

There's a command `airflow dags delete` to delete runs that belongs to dag.

> 
> Also what is the difference between [`delete`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/xcom/index.html#airflow.models.xcom.BaseXCom.delete) and `clear`. When is `delete` called?

I think `delete` could be used to delete a specfic run, and `clear` deletes all runs of a dag.

> but if we deleted it already, then it wont be there anymore.

I think you could assume the previous step will be triggered to generate (query from the underlying database) the data.
